### PR TITLE
Google auth

### PR DESCRIPTION
**Pull request by gzamudio**
### Summary
Added:

- Google authentication
- Improves on identity auth
- ABM Roles
- Role Manager (allows to add and remove roles from users)
- Change password to authenticated user.

### CODE REVIEW

File: Controllers/AccountController.cs [hunk 1]
```diff
@@ -7,30 +7,38 @@
...
 using System.Threading.Tasks;
 using System.Linq;
+using System.Security.Claims;
+using Microsoft.AspNetCore.Authorization;
+using System;
+using Microsoft.Extensions.Options;
 
 namespace NetCoreBootstrap.Controllers
-{
+{
```
Comment (gussiciliano) [inline Controllers/AccountController.cs]: remove extra spaces

File: Controllers/AccountController.cs [hunk 2]
```diff
@@ -7,30 +7,38 @@
...
+{    
     [Route("[controller]")]
     public class AccountController : Controller
     {
         private readonly UserManager<User> _userManager;
         private readonly SignInManager<User> _signInManager;
+		private readonly IOptions<IdentityCookieOptions> _identityCookieOptions;
 
-        public AccountController(UserManager<User> userManager, SignInManager<User> signInManager)
+		public AccountController(UserManager<User> userManager, SignInManager<User> signInManager, IOptions<IdentityCookieOptions> identityCookieOptions)
```
Comment (gussiciliano) [inline Controllers/AccountController.cs]: remove extra spaces

File: Views/UserManagement/_UserRoles.cshtml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+	{  
+		@foreach(var role in Model.Roles)
+		{
+			<div>
+				@Html.CheckBoxFor(model => model.Roles[role.Key], new {})
+				@Html.LabelFor(model => model.Roles[role.Key])
+			</div>
+		}
+		@Html.HiddenFor(model => model.SelectedUserId)
+		<div class='form-group'>
```
Comment (gussiciliano) [inline Views/UserManagement/_UserRoles.cshtml]: ident this div block

File: Views/UserManagement/_UserRoles.cshtml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        	<input type='submit' value='Save changes'/>
```
Comment (gussiciliano) [inline Views/UserManagement/_UserRoles.cshtml]: use localization for 'Save changes'

File: Views/UserManagement/_EditRoleForm.cshtml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+@model NetCoreBootstrap.Models.Views.UserManagementViewModel
+
+<div class='form-group'>
+    @Html.DisplayFor(model => model.Name)
+    @Html.EditorFor(model => model.Name)
+	@Html.ValidationMessageFor(model => model.Name, "", new { @class = "text-danger"})
+</div>
```
Comment (gussiciliano) [inline Views/UserManagement/_EditRoleForm.cshtml]: add final line

File: Views/UserManagement/_EditRoleForm.cshtml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+@model NetCoreBootstrap.Models.Views.UserManagementViewModel
+
+<div class='form-group'>
+    @Html.DisplayFor(model => model.Name)
+    @Html.EditorFor(model => model.Name)
+	@Html.ValidationMessageFor(model => model.Name, "", new { @class = "text-danger"})
```
Comment (gussiciliano) [inline Views/UserManagement/_EditRoleForm.cshtml]: ident this

File: Views/UserManagement/Roles.cshtml [hunk 7]
```diff
@@ -1,17 +1,24 @@
 @model NetCoreBootstrap.Models.Views.UserManagementViewModel
 
 <h3>Roles</h3>
+@Html.ActionLink("New Role", "CreateRole", "UserManagement")
+<span>@ViewData["Message"]</span>
 <table>
     <thead>
         <tr>
             <th>Name</th>
```
Comment (gussiciliano) [inline Views/UserManagement/Roles.cshtml]: use localization

File: Views/UserManagement/Roles.cshtml [hunk 8]
```diff
@@ -1,17 +1,24 @@
...
+            <th>Actions</th>
```
Comment (gussiciliano) [inline Views/UserManagement/Roles.cshtml]: use localization

File: Views/UserManagement/Roles.cshtml [hunk 9]
```diff
@@ -1,17 +1,24 @@
 @model NetCoreBootstrap.Models.Views.UserManagementViewModel
 
 <h3>Roles</h3>
+@Html.ActionLink("New Role", "CreateRole", "UserManagement")
```
Comment (gussiciliano) [inline Views/UserManagement/Roles.cshtml]: can you use localization for 'New Role'?

File: Views/UserManagement/Roles.cshtml [hunk 10]
```diff
@@ -1,17 +1,24 @@
 @model NetCoreBootstrap.Models.Views.UserManagementViewModel
 
 <h3>Roles</h3>
```
Comment (gussiciliano) [inline Views/UserManagement/Roles.cshtml]: use localization

File: Views/UserManagement/RoleManager.cshtml [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+	<h4>Users</h4>
+	@Html.ListBox("UserList", Model.UsersListItem, new { @class = "user-list" })
+</div>
+
+<div class='user-roles'>
+	<h4>Roles</h4>
+	@{ Html.RenderPartial("_UserRoles", Model); }
+</div>
+
+<script type='text/javascript'>
```
Comment (gussiciliano) [inline Views/UserManagement/RoleManager.cshtml]: put this in a js file for auth

File: Views/UserManagement/RoleManager.cshtml [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+$(function() 
+{
+	$(".user-list").on('change', function(){
+		$(".user-roles").load('http://localhost:5000/UserManagement/ViewRoles?userId=' + $(this).val());
```
Comment (gussiciliano) [inline Views/UserManagement/RoleManager.cshtml]: un hardcode this direction, is important dont hardcode the host

File: Views/UserManagement/RoleManager.cshtml [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+	});
+});
+</script>
```
Comment (gussiciliano) [inline Views/UserManagement/RoleManager.cshtml]: add final line

File: Views/UserManagement/RoleManager.cshtml [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+<h3>Role manager</h3>
+
+<div>
+	<h4>Users</h4>
+	@Html.ListBox("UserList", Model.UsersListItem, new { @class = "user-list" })
+</div>
+
+<div class='user-roles'>
+	<h4>Roles</h4>
```
Comment (gussiciliano) [inline Views/UserManagement/RoleManager.cshtml]: use localization

File: Views/UserManagement/RoleManager.cshtml [hunk 15]
```diff
@@ -0,0 +1,22 @@
+@model NetCoreBootstrap.Models.Views.RoleManagerViewModel
+
+<h3>Role manager</h3>
```
Comment (gussiciliano) [inline Views/UserManagement/RoleManager.cshtml]: use localization

File: Views/UserManagement/RoleManager.cshtml [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+<div>
+	<h4>Users</h4>
```
Comment (gussiciliano) [inline Views/UserManagement/RoleManager.cshtml]: use localization

File: Views/UserManagement/EditRole.cshtml [hunk 17]
```diff
@@ -0,0 +1,11 @@
+@model NetCoreBootstrap.Models.Views.UserManagementViewModel
+
+<h3>Edit role</h3>
+@using (Html.BeginForm("EditRole", "UserManagement", FormMethod.Post))
+{  
+	@Html.HiddenFor(model => model.RoleId)
+    @Html.Partial("_EditRoleForm", Model);
+    <div class='form-group'>
+        <input type='submit' value='Save changes'/>
```
Comment (gussiciliano) [inline Views/UserManagement/EditRole.cshtml]: use localization

File: Views/UserManagement/EditRole.cshtml [hunk 18]
```diff
@@ -0,0 +1,11 @@
+@model NetCoreBootstrap.Models.Views.UserManagementViewModel
+
+<h3>Edit role</h3>
+@using (Html.BeginForm("EditRole", "UserManagement", FormMethod.Post))
+{  
+	@Html.HiddenFor(model => model.RoleId)
```
Comment (gussiciliano) [inline Views/UserManagement/EditRole.cshtml]: ident this

File: Views/UserManagement/EditRole.cshtml [hunk 19]
```diff
@@ -0,0 +1,11 @@
+@model NetCoreBootstrap.Models.Views.UserManagementViewModel
+
+<h3>Edit role</h3>
```
Comment (gussiciliano) [inline Views/UserManagement/EditRole.cshtml]: use localization

File: Views/UserManagement/CreateRole.cshtml [hunk 20]
```diff
@@ -3,11 +3,7 @@
...
 @using (Html.BeginForm("CreateRole", "UserManagement", FormMethod.Post))
 {  
-    <div class='form-group'>
-        @Html.DisplayFor(model => model.Name)
-        @Html.EditorFor(model => model.Name)
-        @Html.ValidationMessageFor(model => model.Name, "", new { @class = "text-danger"})
-    </div>
+    @Html.Partial("_EditRoleForm", Model);
     <div class='form-group'>
         <input type='submit' value='Create role'/>
```
Comment (gussiciliano) [inline Views/UserManagement/CreateRole.cshtml]: use localization

File: Views/UserManagement/CreateRole.cshtml [hunk 21]
```diff
@@ -3,11 +3,7 @@
 <h3>Create role</h3>
 @using (Html.BeginForm("CreateRole", "UserManagement", FormMethod.Post))
 {  
-    <div class='form-group'>
-        @Html.DisplayFor(model => model.Name)
-        @Html.EditorFor(model => model.Name)
-        @Html.ValidationMessageFor(model => model.Name, "", new { @class = "text-danger"})
-    </div>
+    @Html.Partial("_EditRoleForm", Model);
```
Comment (gussiciliano) [inline Views/UserManagement/CreateRole.cshtml]: Great! 🤙

File: Views/Shared/_UserManagementPartial.cshtml [hunk 22]
```diff
@@ -0,0 +1,10 @@
+@using NetCoreBootstrap.Models.Database
+@using Microsoft.AspNetCore.Identity
+@inject SignInManager<User> SignInManager
+
+@if (SignInManager.IsSignedIn(User))
+{
+	<li><a asp-area="" asp-controller="UserManagement" asp-action="Users">Users</a></li>
```
Comment (gussiciliano) [inline Views/Shared/_UserManagementPartial.cshtml]: use localization

File: Views/Shared/_UserManagementPartial.cshtml [hunk 23]
```diff
@@ -0,0 +1,10 @@
...
+    <li><a asp-area="" asp-controller="UserManagement" asp-action="Roles">Roles</a></li>
```
Comment (gussiciliano) [inline Views/Shared/_UserManagementPartial.cshtml]: use localization

File: Views/Shared/_UserManagementPartial.cshtml [hunk 24]
```diff
@@ -0,0 +1,10 @@
...
+    <li><a asp-area="" asp-controller="UserManagement" asp-action="RoleManager">Role Manager</a></li>
```
Comment (gussiciliano) [inline Views/Shared/_UserManagementPartial.cshtml]: use localization

File: Views/Shared/_UserManagementPartial.cshtml [hunk 25]
```diff
@@ -0,0 +1,10 @@
...
+}
```
Comment (gussiciliano) [inline Views/Shared/_UserManagementPartial.cshtml]: remember to add final line

File: Views/Shared/_UserManagementPartial.cshtml (hunk 22 above)
Comment (gussiciliano) [reply Views/Shared/_UserManagementPartial.cshtml]: ident this

File: Views/Shared/_LoginPartial.cshtml [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+        </ul>
+    </form>
+}
+else
+{
+    <ul class="nav navbar-nav navbar-right">
+        <li><a asp-area="" asp-controller="Account" asp-action="Register">Register</a></li>
+        <li><a asp-area="" asp-controller="Account" asp-action="Login">Log in</a></li>
+    </ul>
+}
```
Comment (gussiciliano) [inline Views/Shared/_LoginPartial.cshtml]: add final lines

File: Views/Shared/_LoginPartial.cshtml [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+            <li>
+                <button type="submit" class="btn btn-link navbar-btn navbar-link">Log out</button>
+            </li>
+        </ul>
+    </form>
+}
+else
+{
+    <ul class="nav navbar-nav navbar-right">
+        <li><a asp-area="" asp-controller="Account" asp-action="Register">Register</a></li>
```
Comment (gussiciliano) [inline Views/Shared/_LoginPartial.cshtml]: use localization

File: Views/Shared/_LoginPartial.cshtml [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+        <li><a asp-area="" asp-controller="Account" asp-action="Login">Log in</a></li>
```
Comment (gussiciliano) [inline Views/Shared/_LoginPartial.cshtml]: use localization

File: Views/Shared/_LoginPartial.cshtml [hunk 29]
```diff
@@ -0,0 +1,26 @@
...
+
+@if (SignInManager.IsSignedIn(User))
+{
+    <form asp-area="" asp-controller="Account" asp-action="Logout" method="post" id="logoutForm" class="navbar-right">
+        <ul class="nav navbar-nav navbar-right">
+            <li>
+                @Html.ActionLink($"Hello {UserManager.GetUserName(User)}!", "Edit", "Account")
+            </li>
+            <li>
+                <button type="submit" class="btn btn-link navbar-btn navbar-link">Log out</button>
```
Comment (gussiciliano) [inline Views/Shared/_LoginPartial.cshtml]: use localization

File: Views/Shared/_LoginPartial.cshtml [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+
+@inject SignInManager<User> SignInManager
+@inject UserManager<User> UserManager
+
+@if (SignInManager.IsSignedIn(User))
+{
+    <form asp-area="" asp-controller="Account" asp-action="Logout" method="post" id="logoutForm" class="navbar-right">
+        <ul class="nav navbar-nav navbar-right">
+            <li>
+                @Html.ActionLink($"Hello {UserManager.GetUserName(User)}!", "Edit", "Account")
```
Comment (gussiciliano) [inline Views/Shared/_LoginPartial.cshtml]: use localization for Hello

File: Views/Account/Login.cshtml [hunk 31]
```diff
@@ -23,3 +26,13 @@
         <input type="submit" value="Login"/>
     </div>
 }
+@{
+    var loginProviders = SignInManager.GetExternalAuthenticationSchemes().ToList();
+    @using (Html.BeginForm("ExternalLogin", "Account", FormMethod.Post))
+    {
+        foreach (var provider in loginProviders)
+        {
+            <button type="submit" class="btn btn-default" name="provider" value="@provider.AuthenticationScheme" title="Log in using your @provider.DisplayName account">@provider.AuthenticationScheme</button>
```
Comment (gussiciliano) [inline Views/Account/Login.cshtml]: Great! 🥇

Comment (gussiciliano) [reply Views/Account/Login.cshtml]: use localization

File: Views/Account/Edit.cshtml [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+    <div class='form-group'>
+        @Html.LabelFor(model => model.ConfirmNewPassword)
+        @Html.EditorFor(model => model.ConfirmNewPassword)
+        @Html.ValidationMessageFor(model => model.ConfirmNewPassword, "", new { @class = "text-danger"})
+    </div>
+
+    <div class='form-group'>
+        <input type='submit' value='Save changes'/>
+    </div>
+}
```
Comment (gussiciliano) [inline Views/Account/Edit.cshtml]: add final line

File: Views/Account/Edit.cshtml [hunk 33]
```diff
@@ -0,0 +1,30 @@
...
+    </div>
+
+    <div class='form-group'>
+        @Html.LabelFor(model => model.ConfirmNewPassword)
+        @Html.EditorFor(model => model.ConfirmNewPassword)
+        @Html.ValidationMessageFor(model => model.ConfirmNewPassword, "", new { @class = "text-danger"})
+    </div>
+
+    <div class='form-group'>
+        <input type='submit' value='Save changes'/>
```
Comment (gussiciliano) [inline Views/Account/Edit.cshtml]: use localization

File: Views/Account/Edit.cshtml [hunk 34]
```diff
@@ -0,0 +1,30 @@
+@model NetCoreBootstrap.Models.Views.UserManagementViewModel
+
+<h3>Profile</h3>
```
Comment (gussiciliano) [inline Views/Account/Edit.cshtml]: use localization

File: Views/Account/Edit.cshtml [hunk 35]
```diff
@@ -0,0 +1,30 @@
...
+<h4>Change password</h4>
```
Comment (gussiciliano) [inline Views/Account/Edit.cshtml]: use localization

File: Views/Account/Edit.cshtml [hunk 36]
```diff
@@ -0,0 +1,30 @@
...
+@using (Html.BeginForm("EditUser", "Account", FormMethod.Post))
+{
+    <div asp-validation-summary="All"></div>
+    @Html.HiddenFor(model => model.UserId)
+    <div class='form-group'>
+        @Html.LabelFor(model => model.Password)
+        @Html.EditorFor(model => model.Password)
+        @Html.ValidationMessageFor(model => model.Password, "", new { @class = "text-danger"})
+    </div>
+
```
Comment (gussiciliano) [inline Views/Account/Edit.cshtml]: remove this line

File: Views/Account/Edit.cshtml [hunk 37]
```diff
@@ -0,0 +1,30 @@
...
+    <div class='form-group'>
+        @Html.LabelFor(model => model.NewPassword)
+        @Html.EditorFor(model => model.NewPassword)
+        @Html.ValidationMessageFor(model => model.NewPassword, "", new { @class = "text-danger"})
+    </div>
+
```
Comment (gussiciliano) [inline Views/Account/Edit.cshtml]: remove this line

File: Views/Account/Edit.cshtml [hunk 38]
```diff
@@ -0,0 +1,30 @@
...
+    <div class='form-group'>
+        @Html.LabelFor(model => model.ConfirmNewPassword)
+        @Html.EditorFor(model => model.ConfirmNewPassword)
+        @Html.ValidationMessageFor(model => model.ConfirmNewPassword, "", new { @class = "text-danger"})
+    </div>
+
```
Comment (gussiciliano) [inline Views/Account/Edit.cshtml]: remove this line

File: Views/Account/AccessDenied.cshtml [hunk 39]
```diff
@@ -1 +1,8 @@
-<h1>Access Denied</h1>
+@{
+    ViewData["Title"] = "Access Denied";
+}
+
+<header>
+    <h1 class="text-danger">Access Denied.</h1>
```
Comment (gussiciliano) [inline Views/Account/AccessDenied.cshtml]: use localization

File: Views/Account/AccessDenied.cshtml [hunk 40]
```diff
@@ -1 +1,8 @@
...
+    <p class="text-danger">You do not have access to this resource.</p>
```
Comment (gussiciliano) [inline Views/Account/AccessDenied.cshtml]: use localization

File: Views/Account/AccessDenied.cshtml [hunk 41]
```diff
@@ -1 +1,8 @@
-<h1>Access Denied</h1>
+@{
+    ViewData["Title"] = "Access Denied";
```
Comment (gussiciliano) [inline Views/Account/AccessDenied.cshtml]: use localization

File: Startup.cs [hunk 42]
```diff
@@ -73,6 +74,11 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerF
 
             app.UseStaticFiles();
             app.UseIdentity();
+            app.UseGoogleAuthentication(new GoogleOptions()
```
Comment (gussiciliano) [inline Startup.cs]: are we added this for init script?

File: Startup.cs [hunk 43]
```diff
@@ -47,6 +47,7 @@ public void ConfigureServices(IServiceCollection services)
             services.AddIdentity<User, IdentityRole>(options => {
                                                                     options.Cookies.ApplicationCookie.LoginPath = "/Account/Login";
                                                                     options.Cookies.ApplicationCookie.AccessDeniedPath = "/Account/AccessDenied";
+                                                                    options.User.RequireUniqueEmail = true;
```
Comment (gussiciliano) [inline Startup.cs]: are we added this for init script?

File: Repositories/UserRepository.cs [hunk 44]
```diff
@@ -64,10 +64,71 @@ public List<SelectListItem> GetRoles()
             return await UserManager.AddToRoleAsync(user, role);
```
Comment (gussiciliano) [inline Repositories/UserRepository.cs]: ident class with their methods

File: Models/Views/UserManagementViewModel.cs [hunk 45]
```diff
@@ -1,18 +1,31 @@
...
         public List<IdentityRole> Roles { get; set; }
+
+        [Display(Name = "Current Password"), DataType(DataType.Password)]
+        public string Password { get; set; }
+
+        [MinLength(6), MaxLength(40), DataType(DataType.Password), Display(Name = "New password")]
+        public string NewPassword { get; set; }
+     
+        [MinLength(6), MaxLength(40), DataType(DataType.Password), Display(Name = "Confirm new password")]
+        [Compare("NewPassword", ErrorMessage = "The password doesn't match the confirmation password")]
```
Comment (gussiciliano) [inline Models/Views/UserManagementViewModel.cs]: use localization

File: Models/Views/UserManagementViewModel.cs [hunk 46]
```diff
@@ -1,18 +1,31 @@
...
         public List<User> Users { get; set; }
         public List<IdentityRole> Roles { get; set; }
+
+        [Display(Name = "Current Password"), DataType(DataType.Password)]
+        public string Password { get; set; }
+
+        [MinLength(6), MaxLength(40), DataType(DataType.Password), Display(Name = "New password")]
+        public string NewPassword { get; set; }
+     
+        [MinLength(6), MaxLength(40), DataType(DataType.Password), Display(Name = "Confirm new password")]
```
Comment (gussiciliano) [inline Models/Views/UserManagementViewModel.cs]: use localization

File: Models/Views/UserManagementViewModel.cs [hunk 47]
```diff
@@ -1,18 +1,31 @@
...
         public string Email { get; set; }
         public string NewRole { get; set; }
         public List<SelectListItem> RolesListItem { get; set; }
         public List<User> Users { get; set; }
         public List<IdentityRole> Roles { get; set; }
+
+        [Display(Name = "Current Password"), DataType(DataType.Password)]
+        public string Password { get; set; }
+
+        [MinLength(6), MaxLength(40), DataType(DataType.Password), Display(Name = "New password")]
```
Comment (gussiciliano) [inline Models/Views/UserManagementViewModel.cs]: use localization

File: Controllers/UserManagementController.cs [hunk 48]
```diff
@@ -72,6 +72,77 @@ public UserManagementController(DbContextOptions<DataBaseContext> options, UserM
...
+            var userRoles = await UserRepository.GetRoles(user);
+            bool isCurrentRole = false;
+            foreach(var role in viewModel.Roles)
+            {
+                isCurrentRole = userRoles.Contains(role.Key);
+                if(isCurrentRole)
+                {
+                    if(!role.Value) await UserRepository.RemoveRoleFromUser(user, role.Key);
+                }
+                else
```
Comment (gussiciliano) [inline Controllers/UserManagementController.cs]: use else if

File: Controllers/UserManagementController.cs [hunk 49]
```diff
@@ -72,6 +72,77 @@ public UserManagementController(DbContextOptions<DataBaseContext> options, UserM
...
+            var result = await UserRepository.UpdateRole(viewModel.RoleId, viewModel.Name);
+            if(result) ViewData["Message"] = "The role was successfully updated.";
+            else ViewData["Message"] = "The role could not be updated.";
+            return View("./Views/UserManagement/Roles.cshtml", new UserManagementViewModel { Roles = UserRepository.GetAllRoles() });
+        }
+
+        [HttpGet("RoleManager")]
+        public IActionResult RoleManager()
+        {
+            var viewModel = new RoleManagerViewModel { UsersListItem = UserRepository.GetUsersListItem(), Roles = new Dictionary<string, bool>() };
```
Comment (gussiciliano) [inline Controllers/UserManagementController.cs]: use one line

File: Controllers/UserManagementController.cs [hunk 50]
```diff
@@ -72,6 +72,77 @@ public UserManagementController(DbContextOptions<DataBaseContext> options, UserM
...
+            var role = await UserRepository.GetRoleById(roleId);
+            var roleViewModel = new UserManagementViewModel { RoleId = role.Id, Name = role.Name };
+            return View(roleViewModel);
+        }
+
+        [HttpPost("EditRole")]
+        public async Task<IActionResult> EditRole(UserManagementViewModel viewModel)
+        {
+            var result = await UserRepository.UpdateRole(viewModel.RoleId, viewModel.Name);
+            if(result) ViewData["Message"] = "The role was successfully updated.";
```
Comment (gussiciliano) [inline Controllers/UserManagementController.cs]: use localization

File: Controllers/UserManagementController.cs [hunk 51]
```diff
@@ -72,6 +72,77 @@ public UserManagementController(DbContextOptions<DataBaseContext> options, UserM
...
+            else ViewData["Message"] = "The role could not be updated.";
```
Comment (gussiciliano) [inline Controllers/UserManagementController.cs]: use localization

File: Controllers/UserManagementController.cs [hunk 52]
```diff
@@ -72,6 +72,77 @@ public UserManagementController(DbContextOptions<DataBaseContext> options, UserM
             return View(viewModel);
         }
 
+        [HttpGet("DeleteRole")]
+        public async Task<IActionResult> DeleteRole(string roleId)
+        {
+            var result = await UserRepository.DeleteRole(roleId);
+            if(result) ViewData["Message"] = "The role was successfully deleted.";
```
Comment (gussiciliano) [inline Controllers/UserManagementController.cs]: use localization

File: Controllers/UserManagementController.cs [hunk 53]
```diff
@@ -72,6 +72,77 @@ public UserManagementController(DbContextOptions<DataBaseContext> options, UserM
...
+            else ViewData["Message"] = "The role could not be deleted.";
```
Comment (gussiciliano) [inline Controllers/UserManagementController.cs]: use localization

File: Controllers/AccountController.cs (hunk 2 above)
Comment (gzamudio) [reply Controllers/AccountController.cs]: spaces and tabs

File: Views/UserManagement/_EditRoleForm.cshtml (hunk 6 above)
Comment (gzamudio) [reply Views/UserManagement/_EditRoleForm.cshtml]: spaces and tabs

File: Views/UserManagement/EditRole.cshtml (hunk 18 above)
Comment (gzamudio) [reply Views/UserManagement/EditRole.cshtml]: spaces and tabs

File: Startup.cs (hunk 43 above)
Comment (gzamudio) [reply Startup.cs]: no but it's not necessary

File: Repositories/UserRepository.cs (hunk 44 above)
Comment (gzamudio) [reply Repositories/UserRepository.cs]: spaces and tabs

File: Views/UserManagement/_UserRoles.cshtml (hunk 3 above)
Comment (gzamudio) [reply Views/UserManagement/_UserRoles.cshtml]: spaces and tabs

File: Views/UserManagement/_EditRoleForm.cshtml (hunk 5 above)
Comment (gzamudio) [reply Views/UserManagement/_EditRoleForm.cshtml]: ok

File: Startup.cs (hunk 42 above)
Comment (gzamudio) [reply Startup.cs]: yes, can you open an issue for this?

File: Startup.cs (hunk 43 above)
Comment (gussiciliano) [reply Startup.cs]: Oks

File: Startup.cs (hunk 42 above)
Comment (gussiciliano) [reply Startup.cs]: Yes, im going to add right now :)

File: Models/Views/UserManagementViewModel.cs (hunk 46 above)
Comment (gzamudio) [reply Models/Views/UserManagementViewModel.cs]: yes, can you open an issue for this?

File: Models/Views/UserManagementViewModel.cs (hunk 47 above)
Comment (gzamudio) [reply Models/Views/UserManagementViewModel.cs]: yes, can you open an issue for this?

File: Models/Views/UserManagementViewModel.cs (hunk 45 above)
Comment (gzamudio) [reply Models/Views/UserManagementViewModel.cs]: yes, can you open an issue for this?

File: Models/Views/UserManagementViewModel.cs (hunk 47 above)
Comment (gussiciliano) [reply Models/Views/UserManagementViewModel.cs]: done

File: Models/Views/UserManagementViewModel.cs (hunk 46 above)
Comment (gussiciliano) [reply Models/Views/UserManagementViewModel.cs]: done

File: Models/Views/UserManagementViewModel.cs (hunk 45 above)
Comment (gussiciliano) [reply Models/Views/UserManagementViewModel.cs]: done

File: Repositories/UserRepository.cs (hunk 44 above)
Comment (gussiciliano) [reply Repositories/UserRepository.cs]: Oks

File: Views/UserManagement/_EditRoleForm.cshtml (hunk 6 above)
Comment (gussiciliano) [reply Views/UserManagement/_EditRoleForm.cshtml]: Oks

File: Views/UserManagement/_UserRoles.cshtml (hunk 3 above)
Comment (gussiciliano) [reply Views/UserManagement/_UserRoles.cshtml]: Oks

File: Controllers/AccountController.cs [hunk 54]
```diff
@@ -57,6 +71,103 @@ public AccountController(UserManager<User> userManager, SignInManager<User> sign
...
+            }
+
+            // Sign in the user with this external login provider if the user already has a login.
+            var result = await SignInManager.ExternalLoginSignInAsync(info.LoginProvider, info.ProviderKey, isPersistent: false);
+            if (result.Succeeded)
+            {
+                return RedirectToAction("Users", "UserManagement");
+            }
+            else if(result.IsLockedOut || result.IsNotAllowed || result.RequiresTwoFactor) throw new Exception(result.ToString());
+            else
```
Comment (gussiciliano) [inline Controllers/AccountController.cs]: use else if

**Pull request merged**